src/ctap/data_formats.rs

```diff
@@ -432,6 +440,40 @@ impl TryFrom<&cbor::Value> for SignatureAlgorithm {
     }
 }
 
+#[derive(Clone, Copy, PartialEq, PartialOrd)]
+#[cfg_attr(any(test, feature = "debug_ctap"), derive(Debug))]
+pub enum CredentialProtectionPolicy {
+    UserVerificationOptional = 0x01,
+    UserVerificationOptionalWithCredentialIdList = 0x02,
+    UserVerificationRequired = 0x03,
+}
+
+impl TryFrom<cbor::Value> for CredentialProtectionPolicy {
```

Comment:
If it is a short term plan to use move semantics for parsing, then you can disregard everything below.

If we want to avoid this code duplication, we can either:
- Remove this implementation and change line 559
    - from `.map(CredentialProtectionPolicy::try_from)`
    - to `.map(|x| CredentialProtectionPolicy::try_from(&x))`
- Implement it using the call-by-reference version:
    ```
    fn try_from(cbor_value: cbor::Value) -> Result<Self, Ctap2StatusCode> {
        Self::try_from(&cbor_value)
    }
    ```

We can also remove the `extract_integer` function because CBOR integers are essentially `Copy` so it's fine to take them as reference and return a copy of the contained integer. The `extract_` functions are only useful for non-`Copy` types (i.e. move-only types). However, if one day we only use move semantics for parsing, then we can as well use `extract_` and call-by-value for all methods and have a single implementation for each.

Reply:
Plan is to make everything

- extract_*
- try_from without reference

as you suggested and started. So I leave it as is for now?

Reply:
Yes, if it's planned to complete the migration before the formats are modified there's no problem. The risk with code duplication would be that another change needs to update some struct or enum and forgets to update all duplicated parts.